### DISCUSSION TITLE
More buttons on Toolbar

### DISCUSSION THREAD

**Discussion by twicemice** [TriliumNext/Trilium#3298] [Q&A] [answered]
Hi
Is there a way to add more buttons on the toolbar? I have found a limit of 9, and i need more.
or a way to create a button with multiple options, like the "main menu button".

Thank you for your attention.

**Comment by zadam** [comment]
Hi,

there are [bookmarks](https://github.com/zadam/trilium/wiki/Bookmarks) which put shortcuts on the toolbar/panel.

There's also an option to make a custom button using scripts - there's an example on the [wiki](https://github.com/zadam/trilium/wiki/Scripts).

Finally, I have a work-in-progress making the left sidebar much more configurable, but it will take some time to release that.

**Reply by twicemice** [reply]
Thank you, very useful feature but i need that the button has some funcionality, like create new note.

**Comment by fooyoo** [comment, marked as answer]
center-pane can also be utilized(another option).  here is an example for your reference, you can ignore the implementation of button function:
``
const TPL = `
<div>
    <button type="button"
        class="toc-toggle floating-button btn bx bx-font-family no-print"
        title="TOC"></button>
    <button type="button"
        class="pad-toggle floating-button btn bx bx-comment-detail no-print"
        title="scratch"></button>            
    <button type="button"
        class="tbh-toggle floating-button btn bx bx-border-bottom no-print"
        title="border hide"></button>            
    <button type="button"
        class="color-toggle floating-button btn bx bx-border-bottom no-print"
        title="color"></button>                        
    <button type="button"
        class="newline floating-button btn bx bx-subdirectory-left no-print"
        title="color"></button>                                    
</div>`;

class TGLButtonsWidget extends api.NoteContextAwareWidget {
    get position() { return 1; } 
    get parentWidget() { return 'center-pane'; }

    isEnabled() {
        return super.isEnabled() && this.note.type === 'text';
    }

    doRender() {
        super.doRender();
        this.$widget = $(TPL);
        this.$toc = this.$widget.find(".toc-toggle");
        this.$pad = this.$widget.find(".pad-toggle");
        this.$tbh = this.$widget.find(".tbh-toggle");
        this.$color = this.$widget.find(".color-toggle");
        this.$newline = this.$widget.find(".newline");

        this.$toc.on('click', () =>{
                api.runOnBackend(async (noteId) => {
                        const curNote = api.getNote(noteId);
                        if (!curNote.getLabelValue("toc")) {
                        curNote.addLabel("toc", "show");
                        } else if (curNote.getLabelValue("toc") == "show") {
                        curNote.removeLabel("toc");
                        curNote.addLabel("toc", "hide");
                        } else if (curNote.getLabelValue("toc") == "hide") {
                        curNote.removeLabel("toc");
                        }
                        }, [this.note.noteId]);
                });       

        this.$pad.on('click', () =>{
                api.runOnBackend(async (noteId) => {
                        const curNote = api.getNote(noteId);
                        if (!curNote.hasLabel("scratchpadWidgetEnabled")) {
                        curNote.addLabel("scratchpadWidgetEnabled");
                        } else  {
                        curNote.removeLabel("scratchpadWidgetEnabled");
                        }
                        }, [this.note.noteId]);
                });       

        this.$tbh.on('click', () =>{
                api.runOnBackend(async (noteId) => {
                        const note = api.getNote(noteId);
                        const content=note.getContent();
                        const border="border:2px solid var(--main-background-color)";
                        const newcontent=content.replace(/border:2px solid #BADBAD/gm,border);
                        note.removeLabel("TBH");
                        note.setContent(newcontent);
                        note.save();
                        }, [ this.note.noteId]);
                api.waitUntilSynced();
                api.triggerEvent('noteTypeMimeChanged', {noteId: this.note.noteId});             
                }); 

        this.$color.on('click', () =>{
                api.runOnBackend(async (noteId) => {
                        const curNote = api.getNote(noteId);
                        if (!curNote.hasLabel("color")) {
                        curNote.addLabel("color", "orange");
                        } else  {
                        curNote.removeLabel("color");
                        }
                        }, [ this.note.noteId]);
                });       

        this.$newline.on('click', () =>{
                api.runOnBackend(async (noteId) => {
                        const note = api.getNote(noteId);
                        const content=note.getContent();
                        const newcontent=content + "\n<p>&nbsp;</p>\n"
                        note.setContent(newcontent);
                        note.save();
                        }, [ this.note.noteId]);
                api.waitUntilSynced();
                api.triggerEvent('noteTypeMimeChanged', {noteId: this.note.noteId});  
                }); 


        this.contentSized();
        return this.$widget;
    }

}

module.exports = new TGLButtonsWidget();
``
result:
![center-pane-butter](https://user-images.githubusercontent.com/36770100/200786483-7124e1b4-4e5e-441b-8e24-e5315b1643b0.png)

**Reply by fooyoo** [reply]
seems some code is uable to display. paste picture here:
![part-of-code](https://user-images.githubusercontent.com/36770100/200787729-f3de4aa7-34de-49de-bd26-6bbf39c43826.png)

**Reply by twicemice** [reply]
Thank you very much, i´m going to try to adapt this code, i´m a newbie with javascript.

**Comment by twicemice** [comment]
Thank you @fooyoo for your code and tips.
This is the code I have adapted. It renders a new button and if you click on that, it generates a new child note:

```
const TPL = `
<div>
         
    <button type="button"
        class="search1-toggle floating-button btn bx bx-file-find no-print"
        title="search1"></button>     
        
  
</div>`;

class TGLButtonsWidget extends api.NoteContextAwareWidget {
    get position() { return 1; }
    get parentWidget() { return 'center-pane'; }
    isEnabled(){
        return super.isEnabled() && this.note.type === 'text'&& this.note.hasLabel('buttons2');
    }

    doRender() {
        super.doRender();
        this.$widget = $(TPL);
        this.$search1 = this.$widget.find(".search1-toggle");
       
        
        this.$search1.on('click', () =>{
            api.runOnBackend(async (noteId) => {
                    const curNote = api.getNote(noteId);
                                        
                    const response = api.createNewNote({
                            parentNoteId: noteId,
                            title:"search result",
                            content: "",
                            type:"text",
                            isProtected: false
                    });
                    const templateNote=api.getNotesWithLabel('P_search1')[0];      
                    response.note.setRelation('template', templateNote.noteId);            
                    return response.note;     
                          
                          
                    }, [this.note.noteId]);
            api.waitUntilSynced();
            api.activateNewNote(response);
                                
            });       
                    
        this.contentSized();
        return this.$widget;
    }
}   
module.exports = new TGLButtonsWidget(); 
```

Last questions (i hope ; ) )

I've been trying with the "note-detail-pane" option, but it hasn't worked. What am I doing wrong?
On the other hand, how could I add simple javascript code for execution in a backend script. 
 ```
const Search01 =curNote.getLabelValue("Search");
window.open('https://www.amazon.com/s?k=' + Search01, 'Search01window');
```

Thank you very much for your help. I´m learning a lot.

**Reply by fooyoo** [reply]
I am new to JS/frontend also, let's learn from each other.
I tried to change “center-pane' to "note-detail-pane", it can display as before.
and I tried add your js code snippet to open window, it can be opened, not sure if it your expected result. just paste here for your reference.

```
const TPL = `
<div>
         
    <button type="button"
        class="search1-toggle floating-button btn bx bx-file-find no-print"
        title="search1"></button>     
        
  
</div>`;

class TGLButtonsWidget extends api.NoteContextAwareWidget {
    get position() { return 1; }
    get parentWidget() { return 'note-detail-pane'; }
    isEnabled(){
        return super.isEnabled() && this.note.type === 'text'&& this.note.hasLabel('buttons2');
    }

    doRender() {
        super.doRender();
        this.$widget = $(TPL);
        this.$search1 = this.$widget.find(".search1-toggle");
       
        
        this.$search1.on('click', async () =>{
            const response = await api.runOnBackend(async (noteId) => {
                    const curNote = api.getNote(noteId);
                                        
                    const response = api.createNewNote({
                            parentNoteId: noteId,
                            title:"search result",
                            content: "",
                            type:"text",
                            isProtected: false
                    });
                    const templateNote=api.getNotesWithLabel('P_search1')[0];      
                    response.note.setRelation('template', templateNote.noteId);            
                    return response.note.noteId;     
                          
                          
                    }, [this.note.noteId]);
            api.waitUntilSynced();
            api.activateNewNote(response);
            const Search01 =this.note.getLabelValue("Search");
            window.open('https://www.amazon.com/s?k=' + Search01, 'Search01window');                               
            });       
                    
        this.contentSized();
        return this.$widget;
    }
}   
module.exports = new TGLButtonsWidget(); 
```

result:


![opensearch](https://user-images.githubusercontent.com/36770100/201454699-ca78bb2e-1e91-4626-ab2f-5c76ece0450a.gif)

**Comment by twicemice** [comment]
Hi @fooyoo 
That is exactly the result I was looking for, but unfortunately, the browser window does not open on my system. I copied your code, tried other possibilities but it doesn't work. I will try to see how is the debugging with console to see if it throws some error that will allow me to correct it. 
Thank you very very very much for your help.

**Comment by fooyoo** [comment]
I tried again(copy/paste from last posted), still work. note that script need "JS fronted" note type and label #widget.
search note need label "#buttons2 #Search=abc",  and a template note with label "#P_search1“.
Maybe something is blocked/failed before windows open. You could try debug tools of Trilium, Control+Shift+I to open the dev tool to check warning/error. adding some "print" in code(console.log in frontend and api.log in backend) also helps sometimes(need Control-R to reload frontend to make it effective).  check output of console.log in dev tools' console, and check output of api.log in "menu/show backend log".

**Reply by twicemice** [reply]
I copied  the code again and inspected the console, It was talking about an error with a  "NoteId". Everythig was fine until the new note creation, it never applied the template to this new note. I corrected tags and names from my templates and It worked. Now it applies template and opens web browser. 
This will enhance my workflow a lot.  
Thank you very very very much!
